### ISSUE TITLE
storage: sources can get silently stuck after restart

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#15155]
## Description

Sources can get silently wedged after restart.

We believed this was fixed by https://github.com/MaterializeInc/materialize/pull/15469. It certainly became harder to reproduce. However, there are at least two examples of the same CI failures ("deadline has elapsed") after #15469 was merged:

  * https://github.com/MaterializeInc/materialize/issues/15155#issuecomment-1286159613
  * https://github.com/MaterializeInc/materialize/issues/15026#issuecomment-1286162301

### Zippy `KafkaSources` scenario

The Zippy KafkaSources scenario has started timing out with a `deadline has elapsed` message. The issue is readily reproducible in CI and in scratch. Once it times out, things look like this:

```
materialize=> explain timestamp for select * from  "materialize"."public"."sink8_source";
                              Timestamp                              
---------------------------------------------------------------------
           query timestamp: 1664863001809 (2022-10-04 05:56:41.809) +
                     since:[            0 (1970-01-01 00:00:00.000)]+
                     upper:[            0 (1970-01-01 00:00:00.000)]+
          global timestamp: 1664863001809 (2022-10-04 05:56:41.809) +
   can respond immediately: false                                   +
                                                                    +
 source materialize.public.sink8_source (u17, storage):             +
  read frontier:[            0 (1970-01-01 00:00:00.000)]           +
 write frontier:[            0 (1970-01-01 00:00:00.000)]           +
```

Re-creating the source under a different name while pointing to the same topic causes the data to be properly ingested and be available. So the issue is not about the contents of the topic, but about this particular source.

The log lines that relate to `u17` appear to show that its storaged was killed by the test and then the entire Mz instance was also killed. Maybe this sequence of events is what caused the source to get stuck.

To reproduce:

```
cd test/zippy
while bash -c -e "./mzcompose down -v ; ./mzcompose run default --scenario=KafkaSources" ;  do : ; done
```

### Platform checks `RestartEntireMz` scenario

*Originally filed as #15026.*

This is a genuine bug in the product that affects multiple CI jobs. What is happening is that a random source will get stuck and will be impossible to query:

```
---------------------------------------------------------------------
           query timestamp: 1664358630482 (2022-09-28 09:50:30.482) +
                     since:[1664357060000 (2022-09-28 09:24:20.000)]+
                     upper:[1664357061001 (2022-09-28 09:24:21.001)]+
          global timestamp: 1664358630482 (2022-09-28 09:50:30.482) +
   can respond immediately: false                                   +
                                                                    +
 source materialize.public.format_regex1 (u49, storage):            +
  read frontier:[1664357060000 (2022-09-28 09:24:20.000)]           +
 write frontier:[1664357061001 (2022-09-28 09:24:21.001)]           +
```

As you can see, the since and the upper are both 5 minutes in the past. The issue affects only one source on the instance, and it is a different source each time, the other sources work as expected.

To reproduce:

1. Start a scratch instance of the `ci` type
2. Do
```
cd test/platform-checks
while bash -c -e "./mzcompose down -v ; ./mzcompose run  default --scenario RestartEntireMz" ;  do : ; done
```

and wait until it gets stuck and times out .

**Comment by philip-stoev** [comment]
[gh15155.log.zip](https://github.com/MaterializeInc/materialize/files/9703219/gh15155.log.zip)

This is the docker log file from the entire Zippy run. The stuck source is `u17`  . All other sources appear to work

**Comment by philip-stoev** [comment]
Looks very similar to #15026 except that in this situation since = upper = 0 , which I guess is a significant difference. There are also differences in the logging messages that mention the source.

**Comment by philip-stoev** [comment]
I am starting a full run on this branch with default SIZE 4 . Will report once the CI has made some progress.

**Comment by philip-stoev** [comment]
Please disregard this "
I am starting a full run on this branch with default SIZE 4 . Will report once the CI has made some progress.
" . It was not about this ticket.

**Comment by petrosagg** [comment]
Here are some observations from trying to reproduce this problem and by analysing the logs attached by Philip. First, I was unable to reproduce, both locally and on an AWS machine that is identical to CI, the exact problem of this issue. That is, I never observed sinrce = upper = 0. What I was able to reproduce though was the test failing with the same error (`deadline has elapsed`) and got this EXPLAIN TIMESTAMP on the offending object:

```
materialize=> EXPLAIN TIMESTAMP FOR SELECT * FROM view9;
                              Timestamp
---------------------------------------------------------------------
           query timestamp: 1665491962719 (2022-10-11 12:39:22.719) +
                     since:[1665490190000 (2022-10-11 12:09:50.000)]+
                     upper:[1665490191001 (2022-10-11 12:09:51.001)]+
          global timestamp: 1665491962719 (2022-10-11 12:39:22.719) +
   can respond immediately: false                                   +
                                                                    +
 source materialize.public.view9 (u54, storage):                    +
             read frontier:[1665490190000 (2022-10-11 12:09:50.000)]+
            write frontier:[1665490191001 (2022-10-11 12:09:51.001)]+
```

Here we can see that the view is stuck at `12:09`. I traced the problematic dependent source and it was `u32` which I then used to locate the `storaged` process running on the machine. **There wasn't any**. This explains the outcome because the storage controller has stopped receiving upper updates and there is no `storaged` process writing to the output collection. Looking at the logs there is no panic or error. There used to be a `storaged-u32` process which suddenly stops at `12:09`.


Philip's attached log shows a similar thing, the storaged processes are chugging along, including `u17` (the offending source in his case), until they are killed and environmentd is restarted. When environmentd is restarted (and storage controller goes with it to) all frontiers are initialized at `since = upper = 0` until we hear back upper feedback from `storaged` processes. If you look at that log you'll see that during bootstrap source `u17` is rehydrated:

```
environmentd: 2022-10-03T12:49:03.958894Z  INFO mz_adapter::catalog: create source materialize.public.sink8_source (u17)
```

But, contrary to other sources there is no log of the orchestrator spinning up a process! This is what is normally printed for well behaving sources (in this case `u3`):

```
environmentd: 2022-10-03T12:49:04.998136Z  WARN mz_storage::controller::rehydration: error connecting to storage host, retrying: transport error
environmentd: 2022-10-03T12:49:04.998836Z  INFO mz_orchestrator_process: Launching storage-u3: --workers=1 --controller-listen-addr=127.0.0.1:2103 --internal-http-listen-addr=127.0.0.1:2104 --opentelemetry-resource=storage_id=u3 --log-filter=info --log-prefix=storage-u3 --pid-file-location=/mzdata/storage-u3-0.pid --secrets-reader=process --secrets-reader-process-dir=/mzdata/secrets...
```

The `environmentd` then proceeds and periodically (every 30 seconds) logs this line, which I assume is from the storage controller trying to contact the non-existent `storaged-u17` process:

```
environmentd: 2022-10-04T06:03:16.081659Z  WARN mz_storage::controller::rehydration: error connecting to storage host, retrying: transport error
```

while all other storaged instances log:

```
storage-u51: 2022-10-04T06:02:12.159611Z  INFO mz_storage::sink::kafka: kafka-u51: sending progress for gate ts: 1664863333000
```

**Comment by guswynn** [comment]
Its possible that the process orchestrator is finding a storaged child in a weird state or before its fully shutdown, and erroneously not starting the process (https://github.com/MaterializeInc/materialize/blob/main/src/orchestrator-process/src/lib.rs#L218) , but that would not explain the crane problem, which uses `orchestrator-kubernetes`. kubernetes should be restarting the container itself

**Comment by guswynn** [comment]
i don't see how source = upper = 0 could be possible from not starting a process with the kubernetes orchestrator, especially because, at least when i checked, the storaged process appeared to be stuck doing:
```
2022-10-10T22:01:59.160390Z  INFO mz_persist_client::internal::machine: snapshot sc96d3a3b-6a82-4522-b6c8-9897d3573d1a as of Antichain { elements: [1665408296284] } not yet available for upper Antichain { elements: [0] } retrying in 16s
```

**Comment by benesch** [comment]
We've not seen this since https://github.com/MaterializeInc/materialize/pull/15469 went in. Going to optimistically close.

**Comment by benesch** [comment]
I take that right back. I think I see a failure of `KafkaSources` in the nightly tests, just like Philip described in the original issue, here: https://buildkite.com/materialize/nightlies/builds/1254#0183f003-6946-4441-8b75-80d51458d309. AFAICT this nightly run was on a commit that included #15469.

**Comment by aljoscha** [comment]
are these nightlies using K8s or the process orchestrator? above,  @petrosagg and @guswynn did find a bug where our process orchestrator did not spin up a process, even though it should.

**Comment by benesch** [comment]
I filed https://github.com/MaterializeInc/materialize/issues/15725 about how to improve the process orchestrator to fix this. I think there's a pretty glaring bug in the process orchestrator that would lead to this situation.

**Comment by philip-stoev** [comment]
In https://buildkite.com/materialize/tests/builds/44041#018416b9-232f-4350-a9e9-e67ce8918a81 we have a stuck source without a panic in a test that, as the name implies, restarts the entire Mz instance.

The test does that by performing `c.kill("materialized")` which kills the *entire* container ( as confirmed by `docker events`). A new Mz container is then brought up (new cockroach, new everything except mounted volumes). So I do not think any issues in handling orphan children by the process orchestrator will affect this particular CI job.

**Comment by benesch** [comment]
:bulb: I think that's plausibly still consistent with a bug in the process orchestrator. Killing the container but leaving the volumes in place will leave the PID files around...

**Comment by benesch** [comment]
@philip-stoev got this fixed up sufficiently for CI in #15800. We have https://github.com/MaterializeInc/materialize/issues/15725 to track the long-term improvements for the process orchestrator.
